query/transformation.go

```diff
@@ -0,0 +1,195 @@
+)
+
+// A transform takes the list of values, other parameters, and the timerange of the query.
+type transform func([]float64, []value, api.Timerange) ([]float64, error)
+
+// timerangeScale takes an api.Timerange and computes the "scale" or duration of one sample (in seconds).
+// It is useful for transformations that normalize on time (like derivative or integral).
+func timerangeScale(timerange api.Timerange) float64 {
+	return float64(timerange.Resolution)
+}
+
+// transformTimeseries transforms an individual series (rather than an entire serieslist) taking the same parameters as a transform,
+// but with the serieslist standing in for the simplified []float64 argument.
+func transformTimeseries(series api.Timeseries, transform transform, parameters []value, timerange api.Timerange) (api.Timeseries, error) {
+	values, err := transform(series.Values, parameters, timerange)
+	if err != nil {
+		return api.Timeseries{}, err
+	}
+	return api.Timeseries{
+		Values: values,
+		TagSet: series.TagSet,
+	}, nil
+}
+
+// applyTransform applies the given transform to the entire list of series.
+func ApplyTransform(list api.SeriesList, transform transform, parameters []value) (api.SeriesList, error) {
+	result := api.SeriesList{
+		Series:    make([]api.Timeseries, len(list.Series)),
+		Timerange: list.Timerange,
+		Name:      list.Name,
+	}
+	var err error
+	for i, series := range list.Series {
+		result.Series[i], err = transformTimeseries(series, transform, parameters, list.Timerange)
+		if err != nil {
+			return api.SeriesList{}, err
+		}
+	}
+	return result, nil
+}
+
+// checkParameters is used to make sure that each transform is given the right number of parameters.
+func checkParameters(name string, expected int, parameters []value) error {
+	if len(parameters) != expected {
+		printArgs := append([]value{stringValue("(SeriesList)")}, parameters...)
+		return errors.New(fmt.Sprintf("expected %s to be given %d parameters but was given %d: %+v", name, expected+1, len(parameters)+1, printArgs))
+	}
+	return nil
+}
+
+// transformDerivative estimates the "change per second" between the two samples (scaled consecutive difference)
+func transformDerivative(values []float64, parameters []value, timerange api.Timerange) ([]float64, error) {
+	if err := checkParameters("transform.derivative", 0, parameters); err != nil {
+		return nil, err
+	}
+	result := make([]float64, len(values))
+	for i := range values {
+		if i == 0 {
+			// The first element has 0
+			result[i] = 0
+			continue
+		}
+		// Otherwise, it's the scaled difference
+		result[i] = (values[i] - values[i-1]) / timerangeScale(timerange)
+	}
+	return result, nil
+}
+
+// transformIntegral integrates a series whose values are "X per second" to estimate "total X so far"
+// if the series represents "X in this sampling interval" instead, then you should use transformCumulative.
+func transformIntegral(values []float64, parameters []value, timerange api.Timerange) ([]float64, error) {
```

Comment:
A thought about integral.. This behavior of scaling up each entry only makes sense if things get resampled using average instead of sum. Normally the `transformCumulative` would make sense because all data points in between are accounted for where the resolution boundaries are.